File: cases/taint/csharp/recursive-payload-transform-positive/RecursivePayloadTransformPositive.cs
namespace DataFlowBench;

static class RecursivePayloadTransformPositive
{
    static int dfb_source() // DFB-SOURCE: recursive-payload-transform-input
    {
        return 7;
    }

    static void dfb_sink(int value) { } // DFB-SINK: recursive-payload-transform-sink

    static int Walk(int value, int depth)
    {
        if (depth == 0)
        {
            return value; // DFB-WITNESS: recursive-payload-transform-base
        }

        int next = value + 1;
        int recursive = Walk(next, depth - 1); // DFB-WITNESS: recursive-payload-transform-transfer
        return recursive + 1; // DFB-WITNESS: recursive-payload-transform-compose
    }

    static void Run()
    {
        dfb_sink(Walk(dfb_source(), 3));
    }
}

File: cases/taint/go/recursive-payload-transform-positive/recursive_payload_transform_positive.go
package dataflowbench

func dfb_source() int { // DFB-SOURCE: recursive-payload-transform-input
	return 7
}

func dfb_sink(value int) {} // DFB-SINK: recursive-payload-transform-sink

func walk(value int, depth int) int {
	if depth == 0 {
		return value // DFB-WITNESS: recursive-payload-transform-base
	}
	next := value + 1
	recursive := walk(next, depth-1) // DFB-WITNESS: recursive-payload-transform-transfer
	return recursive + 1             // DFB-WITNESS: recursive-payload-transform-compose
}

func run() {
	dfb_sink(walk(dfb_source(), 3))
}
